packages/webdriverio/src/utils.js

```diff
@@ -86,50 +86,54 @@ export const findStrategy = function (value, isW3C) {
         using = 'name'
         value = value.match(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/)[2]
 
-    // any element with given text e.g. h1=Welcome
-    } else if (value.search(/^[a-z0-9]*=(.)+$/) >= 0) {
-        let query = value.split(/=/)
-        let tag = query.shift()
-
-        using = 'xpath'
-        value = `.//${tag.length ? tag : '*'}[normalize-space() = "${query.join('=')}"]`
-
-    // any element containing given text
-    } else if (value.search(/^[a-z0-9]*\*=(.)+$/) >= 0) {
-        let query = value.split(/\*=/)
-        let tag = query.shift()
-
-        using = 'xpath'
-        value = `.//${tag.length ? tag : '*'}[contains(., "${query.join('*=')}")]`
-
-    // any element with certain class or id + given content
-    } else if (value.search(/^[a-z0-9]*(\.|#)-?[_a-zA-Z]+[_a-zA-Z0-9-]*=(.)+$/) >= 0) {
-        let query = value.split(/=/)
-        let tag = query.shift()
-
-        let classOrId = tag.substr(tag.search(/(\.|#)/), 1) === '#' ? 'id' : 'class'
-        let classOrIdName = tag.slice(tag.search(/(\.|#)/) + 1)
-
-        tag = tag.substr(0, tag.search(/(\.|#)/))
-        using = 'xpath'
-        value = `.//${tag.length ? tag : '*'}[contains(@${classOrId}, "${classOrIdName}") and normalize-space() = "${query.join('=')}"]`
-
-    // any element with certain class or id + has certain content
-    } else if (value.search(/^[a-z0-9]*(\.|#)-?[_a-zA-Z]+[_a-zA-Z0-9-]*\*=(.)+$/) >= 0) {
-        let query = value.split(/\*=/)
-        let tag = query.shift()
-
-        let classOrId = tag.substr(tag.search(/(\.|#)/), 1) === '#' ? 'id' : 'class'
-        let classOrIdName = tag.slice(tag.search(/(\.|#)/) + 1)
-
-        tag = tag.substr(0, tag.search(/(\.|#)/))
-        using = 'xpath'
-        value = './/' + (tag.length ? tag : '*') + '[contains(@' + classOrId + ', "' + classOrIdName + '") and contains(., "' + query.join('*=') + '")]'
-        value = `.//${tag.length ? tag : '*'}[contains(@${classOrId}, "${classOrIdName}") and contains(., "${query.join('*=')}")]`
 
     // allow to move up to the parent or select current element
     } else if (value === '..' || value === '.') {
         using = 'xpath'
+
+    // any element with given class, id, or attribute and content
+    // e.g. h1.header=Welcome or [data-name=table-row]=Item or #content*=Intro
+    } else {
+        const match = value.match(new RegExp([
+            // HTML tag
+            /^([a-z0-9]*)/,
+            // optional . or # + class or id
+            /(?:(\.|#)(-?[_a-zA-Z]+[_a-zA-Z0-9-]*))?/,
+            // optional [attribute-name="attribute-value"]
+            /(?:\[(-?[_a-zA-Z]+[_a-zA-Z0-9-]*)(?:=(?:"|')([a-zA-z0-9\-_. ]+)(?:"|'))?\])?/,
+            // *=query or =query
+            /(\*)?=(.+)$/,
+        ].map(rx => rx.source).join('')))
+
+        if (match) {
```

Comment:
If match is `null` then `using` would be `css selector` and value something odd not useful. I would propose to throw an error if no proper using/value pair was found.

Reply:
@christian-bromann please consider that is an inherited behavior, not changing with this PR.

None of these conditions matching would be equivalent to `match` being `null` in the proposed changes:
https://github.com/webdriverio/webdriverio/blob/5a94587926998abed8b78418573b67df712061fe/packages/webdriverio/src/utils.js#L98
https://github.com/webdriverio/webdriverio/blob/5a94587926998abed8b78418573b67df712061fe/packages/webdriverio/src/utils.js#L106
https://github.com/webdriverio/webdriverio/blob/5a94587926998abed8b78418573b67df712061fe/packages/webdriverio/src/utils.js#L118

I think that throwing when `match` is `null` would break the CSS selector use case for `findStrategy` - e.g.`findStrategy('#id')`. Unit tests confirmed locally that thought for me. What do you think?

Reply:
makes sense

Reply:
Thanks! Is there any chance of having this for `v4`?

Reply:
You need to make a separate PR to the v4 branch